lib/Runtime/Library/MathLibrary.cpp

```diff
@@ -714,35 +716,56 @@ namespace Js
             double result = JavascriptConversion::ToNumber(args[1], scriptContext);
             return JavascriptNumber::ToVarNoCheck(result, scriptContext);
         }
-        else if (args.Info.Count == 3)
+        else
         {
-            if (TaggedInt::Is(args[1]) && TaggedInt::Is(args[2]))
+            hasOnlyIntegerArgs = TaggedInt::OnlyContainsTaggedInt(args);
+            if (hasOnlyIntegerArgs && args.Info.Count == 3)
             {
                 return TaggedInt::ToVarUnchecked(max(TaggedInt::ToInt32(args[1]), TaggedInt::ToInt32(args[2])));
             }
         }
 
-        double current = JavascriptConversion::ToNumber(args[1], scriptContext);
-        if(JavascriptNumber::IsNan(current))
+        if (hasOnlyIntegerArgs)
         {
-            return scriptContext->GetLibrary()->GetNaN();
-        }
+            // **Assuming that TaggedInt cannot be NaN?
+            // Int32 vs Int64?
+            int32 current = TaggedInt::ToInt32(args[1]);
+            for (uint idxArg = 2; idxArg < args.Info.Count; idxArg++)
+            {
+                // ** do we need JavascriptNumber::IsNegZero check for TaggedInt?
+                int32 compare = TaggedInt::ToInt32(args[idxArg]);
+                if (current < compare)
+                {
+                    current = compare;
+                }
+            }
 
-        for (uint idxArg = 2; idxArg < args.Info.Count; idxArg++)
+            return TaggedInt::ToVarUnchecked(current);
+        }
+        else
         {
-            double compare = JavascriptConversion::ToNumber(args[idxArg], scriptContext);
-            if(JavascriptNumber::IsNan(compare))
+            double current = JavascriptConversion::ToNumber(args[1], scriptContext);
+            if (JavascriptNumber::IsNan(current))
             {
                 return scriptContext->GetLibrary()->GetNaN();
             }
-            if((JavascriptNumber::IsNegZero(current) && compare == 0) ||
-                current < compare )
+
+            for (uint idxArg = 2; idxArg < args.Info.Count; idxArg++)
             {
-                current = compare;
+                double compare = JavascriptConversion::ToNumber(args[idxArg], scriptContext);
+                if (JavascriptNumber::IsNan(compare))
+                {
+                    return scriptContext->GetLibrary()->GetNaN();
+                }
+                if ((JavascriptNumber::IsNegZero(current) && compare == 0) ||
```

Comment:
Unfortunately, -0.0 === 0.0.  I don't see how we can avoid the check...  It would probably be cheaper to invert the whole thing however:
if (current < compare || (compare == 0 && IsNegZero(current)))

---
In reply to: [133482430](https://github.com/Microsoft/ChakraCore/pull/3495#discussion_r133482430) [](ancestors = 133482430)